libraries/type-length-value/src/state.rs

```diff
@@ -346,25 +453,111 @@ impl<'data> TlvStateMut<'data> {
-    /// the buffer data and zero out the new data.
+    /// Allocate space for the given type, where no repeating discriminators are
+    /// allowed
+    pub fn alloc_unique<V: SplDiscriminate>(
+        &mut self,
+        length: usize,
+    ) -> Result<&mut [u8], ProgramError> {
+        let (bytes, _) = self.alloc::<V>(length, Some(0))?;
+        Ok(bytes)
+    }
+
+    /// Allocate space for the given type, where repeating discriminators _are_
+    /// allowed
+    pub fn alloc_allow_repeating<V: SplDiscriminate>(
+        &mut self,
+        length: usize,
+    ) -> Result<(&mut [u8], usize), ProgramError> {
+        self.alloc::<V>(length, None)
+    }
+
+    /// Allocates and serializes a new TLV entry from a Pod type
+    fn add_entry<V: SplDiscriminate + Pod>(
+        &mut self,
+        value: &V,
+        entry_number: Option<usize>,
+    ) -> Result<usize, ProgramError> {
+        let (data, entry_number) = self.alloc::<V>(size_of::<V>(), entry_number)?;
+        data.copy_from_slice(bytemuck::bytes_of(value));
+        Ok(entry_number)
+    }
+
+    /// Allocates and serializes a new TLV entry from a Pod type, where no
+    /// repeating discriminators are allowed
+    pub fn add_entry_unique<V: SplDiscriminate + Pod>(
+        &mut self,
+        value: &V,
+    ) -> Result<(), ProgramError> {
+        self.add_entry(value, Some(0))?;
+        Ok(())
+    }
+
+    /// Allocates and serializes a new TLV entry from a Pod type, where
+    /// repeating discriminators _are_ allowed. Will add a new entry to the
+    /// next open slot provided there is enough space.
+    pub fn add_entry_allow_repeating<V: SplDiscriminate + Pod>(
+        &mut self,
+        value: &V,
+    ) -> Result<usize, ProgramError> {
+        self.add_entry(value, None)
+    }
+
+    /// Allocates and serializes a new TLV entry from a `VariableLenPack` type
+    fn add_variable_len_entry<V: SplDiscriminate + VariableLenPack>(
+        &mut self,
+        value: &V,
+        entry_number: Option<usize>,
+    ) -> Result<usize, ProgramError> {
+        let length = value.get_packed_len()?;
+        let (data, entry_number) = self.alloc::<V>(length, entry_number)?;
+        value.pack_into_slice(data)?;
+        Ok(entry_number)
+    }
+
+    /// Allocates and serializes a new TLV entry from a `VariableLenPack` type,
+    /// where no repeating discriminators _are_ allowed
+    pub fn add_variable_len_entry_unique<V: SplDiscriminate + VariableLenPack>(
+        &mut self,
+        value: &V,
+    ) -> Result<(), ProgramError> {
+        self.add_variable_len_entry(value, Some(0))?;
+        Ok(())
+    }
+
+    /// Allocates and serializes a new TLV entry from a `VariableLenPack` type,
+    /// where repeating discriminators _are_ allowed. Will add a new entry
+    /// to the next open slot provided there is enough space.
+    pub fn add_variable_len_entry_allow_repeating<V: SplDiscriminate + VariableLenPack>(
```

Comment:
nit naming to make this clearer what's happening
```suggestion
    pub fn alloc_and_pack_variable_len_entry_with_repetition<V: SplDiscriminate + VariableLenPack>(
```